src/storage/v2/indices/text_index.cpp

```diff
@@ -0,0 +1,470 @@
+      throw query::QueryException(fmt::format("Tantivy error: {}", e.what()));
+    }
+  }
+}
+
+void TextIndex::RemoveNode(Vertex *vertex_after_update) {
+  if (!flags::run_time::GetExperimentalTextSearchEnabled()) {
+    throw query::QueryException("To use text indices, enable the text search feature.");
+  }
+
+  auto applicable_text_indices = GetApplicableTextIndices(vertex_after_update);
+  if (applicable_text_indices.empty()) return;
+  RemoveNode(vertex_after_update, applicable_text_indices);
+}
+
+void TextIndex::UpdateOnAddLabel(LabelId added_label, Vertex *vertex_after_update, Storage *storage,
+                                 const std::uint64_t transaction_start_timestamp) {
+  if (!flags::run_time::GetExperimentalTextSearchEnabled()) {
+    throw query::QueryException("To use text indices, enable the text search feature.");
+  }
+
+  if (!label_to_index_.contains(added_label)) {
+    return;
+  }
+  AddNode(vertex_after_update, storage, transaction_start_timestamp,
+          std::vector<mgcxx::text_search::Context *>{&index_.at(label_to_index_.at(added_label)).context_});
+}
+
+void TextIndex::UpdateOnRemoveLabel(LabelId removed_label, Vertex *vertex_after_update,
+                                    const std::uint64_t transaction_start_timestamp) {
+  if (!flags::run_time::GetExperimentalTextSearchEnabled()) {
+    throw query::QueryException("To use text indices, enable the text search feature.");
+  }
+
+  if (!label_to_index_.contains(removed_label)) {
+    return;
+  }
+  RemoveNode(vertex_after_update, {&index_.at(label_to_index_.at(removed_label)).context_});
+}
+
+void TextIndex::UpdateOnSetProperty(Vertex *vertex_after_update, Storage *storage,
+                                    std::uint64_t transaction_start_timestamp) {
+  if (!flags::run_time::GetExperimentalTextSearchEnabled()) {
+    throw query::QueryException("To use text indices, enable the text search feature.");
+  }
+
+  UpdateNode(vertex_after_update, storage, transaction_start_timestamp);
+}
+
+std::vector<mgcxx::text_search::Context *> TextIndex::GetApplicableTextIndices(const std::vector<LabelId> &labels) {
+  if (!flags::run_time::GetExperimentalTextSearchEnabled()) {
+    throw query::QueryException("To use text indices, enable the text search feature.");
+  }
+
+  std::vector<mgcxx::text_search::Context *> applicable_text_indices;
+  for (const auto &label : labels) {
+    if (label_to_index_.contains(label)) {
+      applicable_text_indices.push_back(&index_.at(label_to_index_.at(label)).context_);
+    }
+  }
+  return applicable_text_indices;
+}
+
+std::vector<mgcxx::text_search::Context *> TextIndex::GetApplicableTextIndices(Vertex *vertex) {
+  if (!flags::run_time::GetExperimentalTextSearchEnabled()) {
+    throw query::QueryException("To use text indices, enable the text search feature.");
+  }
+
+  std::vector<mgcxx::text_search::Context *> applicable_text_indices;
+  for (const auto &label : vertex->labels) {
```

Comment:
Don't need to do it here but ranges:filter + transform would fit here quite nicely